contracts/RegistryV1.sol

```diff
@@ -0,0 +1,313 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+import {AccessControlDefaultAdminRulesUpgradeable} from '@openzeppelin/contracts-upgradeable/access/extensions/AccessControlDefaultAdminRulesUpgradeable.sol';
+import {Initializable} from '@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol';
+import {UUPSUpgradeable} from '@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol';
+import {ECDSA} from '@openzeppelin/contracts/utils/cryptography/ECDSA.sol';
+import {MessageHashUtils} from '@openzeppelin/contracts/utils/cryptography/MessageHashUtils.sol';
+
+/**
+ * @title RegistryV1
+ * @notice A registry contract for managing rewards entities and agreements
+ */
+contract RegistryV1 is
+  Initializable,
+  AccessControlDefaultAdminRulesUpgradeable,
+  UUPSUpgradeable
+{
+  using ECDSA for bytes32;
+  using MessageHashUtils for bytes32;
+
+  struct Referral {
+    address user;
+    bytes userSignature;
+    address rewardsConsumer;
+    address rewardsProvider;
+  }
+
+  // Storage
+  uint256 private _nextEntityId; // Counter for generating unique id's
+  mapping(uint256 => address) private _idToEntity; // id => current owner address (Rewards Entity) allows each entity to have a stable identifier while allowing the owner to change
+  mapping(address => uint256) private _entityToId; // current owner => id
```

Comment:
the id management is a bit crazy, but given that the entities are identified by the owner address and we are allowing the transfer of that owner address, i opted to go for this option where the contract knows entities by an internal id that never changes